src/libraries/Hooks.sol

```diff
@@ -151,44 +195,75 @@ library Hooks {
-    ) internal noSelfCall(self) {
-        if (params.liquidityDelta > 0 && self.hasPermission(AFTER_ADD_LIQUIDITY_FLAG)) {
-            self.callHook(
-                abi.encodeWithSelector(IHooks.afterAddLiquidity.selector, msg.sender, key, params, delta, hookData)
-            );
-        } else if (params.liquidityDelta <= 0 && self.hasPermission(AFTER_REMOVE_LIQUIDITY_FLAG)) {
-            self.callHook(
-                abi.encodeWithSelector(IHooks.afterRemoveLiquidity.selector, msg.sender, key, params, delta, hookData)
-            );
+    ) internal noSelfCall(self) returns (BalanceDelta hookDelta) {
+        if (params.liquidityDelta > 0) {
+            if (self.hasPermission(AFTER_ADD_LIQUIDITY_FLAG)) {
+                hookDelta = BalanceDelta.wrap(
+                    self.callHookWithReturnDelta(
+                        abi.encodeWithSelector(
+                            IHooks.afterAddLiquidity.selector, msg.sender, key, params, delta, hookData
+                        ),
+                        self.hasPermission(AFTER_ADD_LIQUIDITY_RETURNS_DELTA_FLAG)
+                    )
+                );
+            }
+        } else {
+            if (self.hasPermission(AFTER_REMOVE_LIQUIDITY_FLAG)) {
+                hookDelta = BalanceDelta.wrap(
+                    self.callHookWithReturnDelta(
+                        abi.encodeWithSelector(
+                            IHooks.afterRemoveLiquidity.selector, msg.sender, key, params, delta, hookData
+                        ),
+                        self.hasPermission(AFTER_REMOVE_LIQUIDITY_RETURNS_DELTA_FLAG)
+                    )
+                );
+            }
         }
     }
 
     /// @notice calls beforeSwap hook if permissioned and validates return value
     function beforeSwap(IHooks self, PoolKey memory key, IPoolManager.SwapParams memory params, bytes calldata hookData)
         internal
-        noSelfCall(self)
+        returns (int256 amountToSwap, int128 hookDeltaSpecified)
     {
+        amountToSwap = params.amountSpecified;
+        if (msg.sender == address(self)) return (amountToSwap, hookDeltaSpecified);
         if (self.hasPermission(BEFORE_SWAP_FLAG)) {
-            self.callHook(abi.encodeWithSelector(IHooks.beforeSwap.selector, msg.sender, key, params, hookData));
+            hookDeltaSpecified = self.callHookWithReturnDelta(
+                abi.encodeWithSelector(IHooks.beforeSwap.selector, msg.sender, key, params, hookData),
+                self.hasPermission(BEFORE_SWAP_RETURNS_DELTA_FLAG)
+            ).toInt128();
+
+            // Update the swap amount according to the hook's return, and check that the swap type doesnt change (exact input/output)
+            if (hookDeltaSpecified != 0) {
+                bool exactInput = amountToSwap < 0;
+                amountToSwap += hookDeltaSpecified;
+                if (exactInput ? amountToSwap > 0 : amountToSwap < 0) revert HookDeltaExceedsSwapAmount();
+            }
```

Comment:
it jumps out a bit to me that this hook in particular is doing a lot of extra legwork that others aren't doing, and especially that a bunch of hook-result legwork is also happening in the top level `poolmanager.swap` function as well. Not sure a specific solution but it feels like the work-split is a bit awkward - would prefer if either the hook calls can handle everything related to hook result parsing, or nothing - currently feels half/half